src/System.Sorting/src/TimSort.cs

```diff
@@ -0,0 +1,680 @@
+        //If the entire array is less than this length, no merges will be performed, instead Binary Insertion sort is used directly.
+        internal const int MinMerge = 32;
+
+        //Most of the merge occurs in what is called "one pair at a time" mode, 
+        //where respective elements of both runs are compared. When the algorithm merges left-to-right, 
+        //the smaller of the two is brought to a merge area. A count of the number of times the final element appears in a given run is recorded. When this value reaches a certain threshold, MIN_GALLOP, the merge switches to "galloping mode".
+        internal const int MinGallop = 7;
+
+        //C uses 85, Java uses a mixed mode of arriving at most efficient number for stackLength. This varies depending on 
+        //MinMerge number above. we may need to readjust these nos, in c# based on code performance.
+        internal const int StackLength = 40;
+
+        //internal T[] mergeBufferArray=new T[256];
+
+        //The minimum run size (minrun) depends on the size of the array.
+        //The final algorithm takes the six most significant bits of the size of the array, adds one if any of the remaining bits are set, and uses that result as the minrun.This algorithm works for all arrays, including those smaller than 32
+        internal static int FindMinimumRun(int arrayLength)
+        {
+            int remainder = 0;  /* becomes 1 if the least significant bits contain at least one off bit */
+            while (arrayLength >= MinMerge)
+            {
+                remainder |= arrayLength & 1;
+                arrayLength >>= 1;
+            }
+            return arrayLength + remainder;
+        }
+
+    }
+    //public class TimSort<T> where T : IComparable<T>
+    public class TimSort<T> where T : IComparable,new()
+    {
+        //counter to increase the arrayRunBase & arrayRunLength
+        private int stackSize;
+
+        /// <summary>
+		/// A stack of pending runs yet to be merged. 
+        /// Run i starts at
+		/// address base[i] and extends for len[i] elements.  It's always
+		/// true (so long as the indices are in bounds) that:
+		/// <c>runBase[i] + runLen[i] == runBase[i + 1]</c>
+        private int[] arrayRunBase;
+        private int[] arrayRunLength;
+
+        /// <summary>
+        /// Temp storage for merges.
+        /// </summary>
+        private T[] mergeBufferArray;// = new T[256];
+        
+        /// <summary>The array being sorted.</summary>
+		private readonly T[] array;
+        private int m_minGallop;
+
+        /*** The comparer for this sort.*/
+        private IComparer<T> vComparable;
+
+        //private TimSort(T[] keys, IComparer<T> comparer)
+        public TimSort(T[] keys, IComparer<T> comparer)
```

Comment:
Hi @ellismg 

the initial array & comparer assignment in the constructor gets used inside functions.
private void MergeAt(int runIndex)
private void MergeLo(int base1, int len1, int base2, int len2)
 private void MergeHi(int base1, int len1, int base2, int len2)

If I do not use this here, I need to carry this to all functions until this gets called to the above functions. Please let me know what you suggest